src/lib.rs

```diff
@@ -250,31 +247,86 @@ impl<TSuccess> From<NonFatalSearchError> for Result<TSuccess, SingleFileSearchEr
     }
 }
 
-pub fn run(args: Args) -> Result<RunStatus, Error> {
-    let query_text = match (args.path_to_query_file.as_ref(), args.query_text.as_ref()) {
-        (Some(path_to_query_file), None) => {
-            fs::read_to_string(path_to_query_file).map_err(|source| Error::QueryFileReadError {
-                source,
-                path_to_query_file: path_to_query_file.clone(),
-            })?
-        }
-        (None, Some(query_text)) => query_text.clone(),
-        (None, None) => ALL_NODES_QUERY.to_owned(),
-        _ => unreachable!(),
-    };
-    let filter =
-        get_loaded_filter(args.filter.as_deref(), args.filter_arg.as_deref())?.map(Arc::new);
+pub struct OutputContext {
+    pub buffer_writer: BufferWriter,
+}
+
+impl OutputContext {
+    pub fn new(buffer_writer: BufferWriter) -> Self {
+        Self { buffer_writer }
+    }
+}
+
+pub fn run_print(args: Args) -> Result<RunStatus, Error> {
+    run_for_context(
+        args,
+        OutputContext::new(BufferWriter::stdout(ColorChoice::Never)),
+        |context: &OutputContext,
+         args: &Args,
+         path: &Path,
+         query_context: QueryContext,
+         matched: &AtomicBool| {
+            let printer = get_printer(&context.buffer_writer, args);
+            let mut printer = printer.borrow_mut();
+
+            printer.get_mut().clear();
+            let mut sink = printer.sink_with_path(path);
+            get_searcher(args)
+                .borrow_mut()
+                .search_path(query_context, path, &mut sink)
+                .unwrap();
+            if sink.has_match() {
+                matched.store(true, Ordering::SeqCst);
+            }
+            context.buffer_writer.print(printer.get_mut()).unwrap();
+        },
+    )
+}
+
+pub fn run_with_callback(
+    args: Args,
+    callback: impl Fn(Node, &[u8], &Path) + Sync,
+) -> Result<RunStatus, Error> {
+    run_for_context(
+        args,
+        (),
+        |_context: &(),
+         args: &Args,
+         path: &Path,
+         query_context: QueryContext,
+         matched: &AtomicBool| {
+            get_searcher(args)
+                .borrow_mut()
+                .search_path_callback::<_, io::Error>(
+                    query_context,
+                    path,
+                    |node: Node, file_contents: &[u8], path: &Path| {
+                        callback(node, file_contents, path);
+                        matched.store(true, Ordering::SeqCst);
+                    },
+                )
+                .unwrap();
```

Comment:
This mimics the `.search_path().unwrap()` above but it's presumably not good to just be `.unwrap()`'ing either one (this would represent eg an `io::Error` while doing the searching)